### DISCUSSION TITLE
Is it intentional that the dev scripts leak memory?

### DISCUSSION THREAD

**Discussion by danieltroger** [parcel-bundler/parcel#8199] [Q&A] [answered]
So I've noticed that when running the dev script  (`yarn serve`) memory is being leaked that isn't when using production builds (`yarn build`).

This seems to happen if you have two files:

file 0.js:
- has a map in the module scope
- exports a function that pushes things into an array
- adds said array to the map

file 1.js (entry point):
- creates 10k <div> nodes
- calls the function from file 0

Then when using the memory profiler in chrome devtools there are 10k detached HTMLDivElement while when doing `parcel build` everything gets GCd

**Comment by danieltroger** [comment]
I'll add that this is annoying because I'm debugging a memory leak and apparently I need to build prod scripts for every iteration while I'm working on it -.-

But maybe it's not worth fixing?

**Comment by mischnic** [comment, marked as answer]
Certainly not "intentional". Either it's a bug or a shortcoming of the dev output format.

Are you talking about HMR? So the old module that was replaced isn't getting GCed?

They should be deleted: https://github.com/parcel-bundler/parcel/blob/a5b43a3c38c39acae2b8e2f83d9d98cf67615539/packages/runtimes/hmr/src/loaders/hmr-runtime.js#L421
But maybe there's a reference to them we forgot to delete

**Reply by danieltroger** [reply]
> a shortcoming of the dev output format.

Yeah that's probably it. It retains a scope for some reason. See attached files.

[parcel_serve_memory_leak.zip](https://github.com/parcel-bundler/parcel/files/8878411/parcel_serve_memory_leak.zip)


<img width="2160" alt="Screenshot 2022-06-10 at 13 12 23" src="https://user-images.githubusercontent.com/4047180/173056033-f0e17f98-cbd8-45d9-b019-a32461afca41.png">
<img width="2160" alt="Screenshot 2022-06-10 at 13 10 49" src="https://user-images.githubusercontent.com/4047180/173056042-8540e49d-89ec-4bcd-93c8-d8b65ef5ca9e.png">

**Reply by mischnic** [reply]
I see. The reason why it's not collected in dev is that it sets `window.parcelRequire94c2` to be a function where you could still access the module via `window.parcelRequire94c2("..the asset id..")`, so the map can't be collected.

But you could run into the same situation even in production builds if the asset with the map is part of a shared bundle or child bundle (= via a dynamic import).

**Reply by danieltroger** [reply]
I see. That doesn't sound optimal?

Btw I'm at the stage where I'll probably need dynamic imports soon but it made the output bundle sooo much uglier (even the names of all the exports were still in there). Is there a way to import an http url with parcel? And to build something that exports something? Because then I could just build solid-js with scope hoisting and import it from the rest of the code with scope hoisting and have more beautiful bundles.

**Reply by mischnic** [reply]
> I see. That doesn't sound optimal?

There's nothing we can do about this. This behaves just like native ESM, so here's a nother another argument for why you shouldn't be relying in scope-hoisting allowing GC:
if you run your code without running Parcel with the browser's native ESM, `map` will also prevent the collection of the elements. But if you change `map` to be a `WeakMap`, then they will be collected in your example

**Reply by danieltroger** [reply]
Mm makes sense. I made it a map with both key and values wrapped in WeakRef because I needed the values to be able to be GCd too. No FinalizationRegistry to delete the WeakRefs because I'm lazy so it will still leak but much less. Thanks
